basesandbox/src/ipc/multiplex.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::{IpcRecv, IpcSend, RecvError, Terminate};
+use crossbeam::channel::bounded;
+use std::thread;
+
+type Sender = crossbeam::channel::Sender<Vec<u8>>;
+type Receiver = crossbeam::channel::Receiver<Vec<u8>>;
+
+pub trait Forward {
+    fn forward(data: &[u8]) -> usize;
+}
+
+fn sender<S: IpcSend>(send: S, recv: Receiver) {
+    loop {
+        let data = recv.recv().unwrap();
+        // special exit flag
+        if data.is_empty() {
+            break
+        }
+        send.send(&data);
+    }
+}
+
+fn receiver<F: Forward, R: IpcRecv>(recv: R, send: Vec<Sender>) {
+    loop {
+        let data = match recv.recv(None) {
+            Err(RecvError::TimeOut) => panic!(),
+            Err(RecvError::Termination) => return,
+            Ok(x) => x,
+        };
+        send[F::forward(&data)].send(data).unwrap();
```

Comment:
I understand the Multiplexer.
IMO, using number index to route packets is not intuitive.

How about this interface? What do you think about it?

```rust
let server_recv;
let client_recv;
Multiplexer::create(ipc_send, ipc_recv, move |data| {
  let header = PacketHeader::new(&data);
  if header.slot >= SLOT_CALL_OR_RETURN_INDICATOR {
    server_recv.send(data);
  } else {
    client_recv.send(data);
  }
}
```

Reply:
I don't get it. Where do you want me to put this code?

Reply:
We talked about it in a call. We concluded that add a unit test to show how to use the Multiplexer for the reader.